server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregationBuilder.java

```diff
@@ -167,38 +217,47 @@ static Rounding createRounding(Rounding.DateTimeUnit interval, ZoneId timeZone)
     @Override
     protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
         builder.field(NUM_BUCKETS_FIELD.getPreferredName(), numBuckets);
+        builder.field(MINIMUM_INTERVAL_FIELD.getPreferredName(), minimumIntervalExpression);
         return builder;
     }
 
     @Override
     protected int innerHashCode() {
-        return Objects.hash(numBuckets);
+        return Objects.hash(numBuckets, minimumIntervalExpression);
     }
 
     @Override
     protected boolean innerEquals(Object obj) {
         AutoDateHistogramAggregationBuilder other = (AutoDateHistogramAggregationBuilder) obj;
-        return Objects.equals(numBuckets, other.numBuckets);
+        return Objects.equals(numBuckets, other.numBuckets) && Objects.equals(minimumIntervalExpression, other.minimumIntervalExpression);
     }
 
     public static class RoundingInfo implements Writeable {
         final Rounding rounding;
         final int[] innerIntervals;
         final long roughEstimateDurationMillis;
         final String unitAbbreviation;
-
-        public RoundingInfo(Rounding rounding, long roughEstimateDurationMillis, String unitAbbreviation, int... innerIntervals) {
-            this.rounding = rounding;
+        final String dateTimeUnit;
+
+        public RoundingInfo(Rounding.DateTimeUnit dateTimeUnit,
+                            ZoneId timeZone,
+                            long roughEstimateDurationMillis,
+                            String unitAbbreviation,
+                            int... innerIntervals) {
+            this.rounding = createRounding(Rounding.DateTimeUnit.SECOND_OF_MINUTE, timeZone);
             this.roughEstimateDurationMillis = roughEstimateDurationMillis;
             this.unitAbbreviation = unitAbbreviation;
             this.innerIntervals = innerIntervals;
+            assert dateTimeUnit != null;
```

Comment:
I think we need to check that the `dateTimeUnit` is not null, but with `Objects.RequireNonNull()`. The assert I suggested was meant for `this.dateTimeUnit` to ensure we never ask for an entry from the map that doesn't exist